File: PolyWorksSDK/samples/applications/IMInspect/VisualBasic.Net/VBNetReadAndFit/MainForm.vb
Public Class MainForm

    ' Global access to a PolyWorks|Inspector
    Private imInspectObject As IMInspectLib.IMInspect = Nothing

    ' Upon loading the Form
    Private Sub MainForm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        imInspectObject = New IMInspectLib.IMInspect

    End Sub

    ' Cleanup upon destruction of the form
    Private Sub MainForm_FormClosed(ByVal sender As System.Object, ByVal e As System.Windows.Forms.FormClosedEventArgs) Handles MyBase.FormClosed

        System.Runtime.InteropServices.Marshal.ReleaseComObject(imInspectObject)
        imInspectObject = Nothing

    End Sub

    ' Main action of this application
    Private Sub ButtonSend_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ButtonSend.Click

        LabelResult.Text = ""

        ' If the server does not exists, return
        If imInspectObject Is Nothing Then

            Exit Sub

        End If

        ' Obtain PolyWorks|Inspector's current project
        Dim projectObject As IMInspectLib.IIMInspectProject = Nothing
        imInspectObject.ProjectGetCurrent(projectObject)

        If projectObject Is Nothing Then

            Exit Sub

        End If

        ' Read the file and fill the coordinates and normals lists
        Dim coordinates As System.Collections.Generic.List(Of Single) = New System.Collections.Generic.List(Of Single)()
        Dim normals As System.Collections.Generic.List(Of Single) = New System.Collections.Generic.List(Of Single)()
        Dim colors As System.Collections.Generic.List(Of Byte) = New System.Collections.Generic.List(Of Byte)()

        If FillArraysFromFile((TextFileName.Text), coordinates, normals) = False Then

            Exit Sub

        End If


        ' Set all points to a blue color
        Dim index As Integer
        index = 0

        While index < coordinates.Count

            colors.Add(32)
            colors.Add(32)
            colors.Add(224)
            index += 3

        End While

        ' Obtain the name of the file, without the path
        Dim tokens As String() = TextFileName.Text.Split(New [Char]() {"\"c})
        Dim objectName As String = tokens(tokens.Length - 1)

        ' Create a point cloud in the current PolyWorks|Inspector project
        Dim pointCloudObject As IMInspectLib.IIMPointCloud
        projectObject.PointCloudCreateData(pointCloudObject, objectName, 0)

        ' Send the points to the point cloud
        If Not pointCloudObject Is Nothing Then

            pointCloudObject.PointsAdd(coordinates.ToArray(), normals.ToArray(), colors.ToArray())

        End If

        ' Obtain a Command Center from the current PolyWorks|Inspector project
        Dim commandCenterObject As IMInspectLib.IIMCommandCenter
        projectObject.CommandCenterCreate(commandCenterObject)

        ' Send commands to fit a sphere on all the elements of this new object
        Dim returnValue As Integer
        Dim script As String
        Dim sphereName As String

        If Not commandCenterObject Is Nothing Then

            commandCenterObject.CommandExecute("TREEVIEW OBJECT SELECT NONE")
            commandCenterObject.CommandExecute("TREEVIEW OBJECT_ELEMENTS SELECT ALL ( """ & objectName & """ )")

            ' Perform the fit and obtain the fitted sphere's name
            script = "version ""5.0""" & vbCrLf &
                         "DECLARE nbSpheres" & vbCrLf &
                         "DECLARE lastSphereName" & vbCrLf &
                         "MACRO END_ON_ERROR ( ""On"" )" & vbCrLf &
                         "FEATURE PRIMITIVE SPHERE FIT_SELECTED_ELEMENTS" & vbCrLf &
                         "TREEVIEW PRIMITIVE SPHERE COUNT GET( nbSpheres )" & vbCrLf &
                         "TREEVIEW PRIMITIVE SPHERE NAME GET( $nbSpheres, lastSphereName )" & vbCrLf

            returnValue = commandCenterObject.ScriptExecuteFromBuffer(script, "")

            If commandCenterObject.ReturnValueIsSuccess(returnValue) = 1 Then

                returnValue = commandCenterObject.ScriptVariableGetValueAsString("lastSphereName", 1, sphereName)

                If commandCenterObject.ReturnValueIsSuccess(returnValue) = 1 Then

                    LabelResult.Text = "Sphere """ & sphereName & """ was successfully fitted"

                End If

            End If

            If commandCenterObject.ReturnValueIsSuccess(returnValue) = 0 Then

                LabelResult.Text = "The fitting was unsuccessful"

            End If

        End If

        ' Cleanup
        System.Runtime.InteropServices.Marshal.ReleaseComObject(pointCloudObject)
        System.Runtime.InteropServices.Marshal.ReleaseComObject(commandCenterObject)
        System.Runtime.InteropServices.Marshal.ReleaseComObject(projectObject)

        pointCloudObject = Nothing
        commandCenterObject = Nothing
        projectObject = Nothing

    End Sub

    ' Reads an ASCII file containing lines made of 6 values (X,Y,Z,I,J,K)
    ' and fills a coordinates and a normals arrays with those values
    Private Function FillArraysFromFile( _
        ByRef filename_ As String, _
        ByRef coordinates_ As System.Collections.Generic.List(Of Single), _
        ByRef normals_ As System.Collections.Generic.List(Of Single)) As Boolean

        ' Validate file name
        If Not System.IO.File.Exists(filename_) Then

            Return False

        End If

        ' Open the file for reading
        Dim reader As New System.IO.StreamReader(filename_)
        Dim line As String
        Dim success As Boolean

        ' Read the first six tokens (X,Y,Z,I,J,K) on each line of the file
        ' and convert them to double-precision floating point values
        success = True
        line = reader.ReadLine()

        While Not line Is Nothing

            Dim tokens As String() = line.Split(New [Char]() {" "c})

            If tokens.Length < 6 Then

                success = False
                Exit While

            End If

            ' Recording coordinates
            coordinates_.Add(Convert.ToSingle(tokens(0)))
            coordinates_.Add(Convert.ToSingle(tokens(1)))
            coordinates_.Add(Convert.ToSingle(tokens(2)))

            ' Recording normals
            normals_.Add(Convert.ToSingle(tokens(3)))
            normals_.Add(Convert.ToSingle(tokens(4)))
            normals_.Add(Convert.ToSingle(tokens(5)))

            line = reader.ReadLine()

        End While

        reader.Close()

        Return success

    End Function

End Class

File: PolyWorksSDK/samples/applications/IMInspect/CSharp/CSharpReadAndFit/MainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.IO;
using System.Text;
using System.Windows.Forms;

namespace CSharpReadAndFit
{
    public partial class MainForm : Form
    {
        // =========================================================================================
        // ========================================= Members =======================================

        private IMInspectLib.IMInspect m_imInspect = null;

        // =========================================================================================
        // ========================================= Methods =======================================

        // =========================================================================================
        // Constructor
        // -----------------------------------------------------------------------------------------
        public MainForm()
        {
            InitializeComponent();
        }

        // =========================================================================================
        // Upon loading the form
        // -----------------------------------------------------------------------------------------
        private void MainForm_Load( object sender, EventArgs e )
        {
            m_imInspect = new IMInspectLib.IMInspect();
        }

        // =========================================================================================
        // Cleanup upon destruction of the form
        // -----------------------------------------------------------------------------------------
        private void MainForm_FormClosed(object sender, FormClosedEventArgs e)
        {
            m_imInspect = null;
        }

        // =========================================================================================
        // Main action of this application
        // -----------------------------------------------------------------------------------------
        private void ButtonSend_Click(object sender, EventArgs e)
        {
            LabelResult.Text = "";

            // If the server does not exists, return
            if ( m_imInspect == null )
            {
                return;
            }

            // Obtain PolyWorks|Inspector's current project
            IMInspectLib.IIMInspectProject imInspectProject = null;
            m_imInspect.ProjectGetCurrent( out imInspectProject );

            if ( imInspectProject == null )
            {
                return;
            }

            // Read the file and fill the coordinates and normals lists
            System.Collections.Generic.List<float> coordinates = new System.Collections.Generic.List<float>();
            System.Collections.Generic.List<float> normals = new System.Collections.Generic.List<float>();
            System.Collections.Generic.List<Byte> colors = new System.Collections.Generic.List<Byte>();

            if ( !FillArraysFromFile( TextFileName.Text, ref coordinates, ref normals ) )
            {
                return;
            }

            // Set all points to a blue color
            for ( int index = 0; index < coordinates.Count; index += 3 )
            {
                colors.Add( 32 );
                colors.Add( 32 );
                colors.Add( 224 );
            }

            // Obtain the name of the file, without the path
            string[] tokens = TextFileName.Text.Split( new Char[] { '\\' } );
            string objectName = tokens[ tokens.Length - 1 ];

            // Create a point cloud in the current PolyWorks|Inspector project
            IMInspectLib.IIMPointCloud imPointCloud = null;
            imInspectProject.PointCloudCreateData( out imPointCloud, objectName, 0 );

            // Send the points to the point cloud
            if ( imPointCloud != null )
            {
                Array coordinatesArray = coordinates.ToArray();
                Array normalsArray = normals.ToArray();
                Array colorsArray = colors.ToArray();

                imPointCloud.PointsAdd( ref coordinatesArray, ref normalsArray, ref colorsArray );
            }

            // Obtain a Command Center from the current PolyWorks|Inspector project
            IMInspectLib.IIMCommandCenter imCommandCenter = null;
            imInspectProject.CommandCenterCreate( out imCommandCenter );

            // Send commands to fit a sphere on all the elements of this new object
            String script;
            String sphereName;

            if ( imCommandCenter != null )
            {
                imCommandCenter.CommandExecute( "TREEVIEW OBJECT SELECT NONE" );
                imCommandCenter.CommandExecute( String.Concat( "TREEVIEW OBJECT_ELEMENTS SELECT ALL ( \"", objectName, "\" ) " ) );

                // Perform the fit and obtain the fitted sphere's name
                script = String.Concat( "version \"5.0\"", "\n", 
                                        "DECLARE nbSpheres", "\n",
                                        "DECLARE lastSphereName", "\n",
                                        "MACRO END_ON_ERROR ( \"On\" )", "\n",
                                        "FEATURE PRIMITIVE SPHERE FIT_SELECTED_ELEMENTS", "\n",
                                        "TREEVIEW PRIMITIVE SPHERE COUNT GET( nbSpheres )", "\n",
                                        "TREEVIEW PRIMITIVE SPHERE NAME GET( $nbSpheres, lastSphereName )", "\n" );

                int returnValue = imCommandCenter.ScriptExecuteFromBuffer( script, "" );
                if ( imCommandCenter.ReturnValueIsSuccess( returnValue ) == 1 )
                {
                    returnValue = imCommandCenter.ScriptVariableGetValueAsString( "lastSphereName", 1, out sphereName );
                    if ( imCommandCenter.ReturnValueIsSuccess( returnValue ) == 1 )
                    {
                        LabelResult.Text = String.Concat( "Sphere \"", sphereName, "\" was successfully fitted" );
                    }
                }
                
                if ( imCommandCenter.ReturnValueIsSuccess( returnValue ) == 0 )
                {                    
                    LabelResult.Text = "The fitting was unsuccessful";
                }
            }

            // Cleanup
            imPointCloud = null;
            imCommandCenter = null;
            imInspectProject = null;
        }

        // =========================================================================================
        // ======================================== Utilities ======================================

        // =========================================================================================
        // Reads an ASCII file containing lines made of 6 values (X,Y,Z,I,J,K)
        // and fills a coordinates and a normals lists with those values
        // -----------------------------------------------------------------------------------------
        private bool FillArraysFromFile(
            String filename_,
            ref System.Collections.Generic.List<float> coordinates_,
            ref System.Collections.Generic.List<float> normals_)
        {
            // Validate file name
            if ( !File.Exists( filename_ ) )
            {
                return false;
            }

            // Open the file for reading
            StreamReader reader = new StreamReader( filename_ );
            String line;
            bool success = true;

            // Read the first six tokens (X,Y,Z,I,J,K) on each line of the file
            // and convert them to double-precision floating point values
            while ( ( line = reader.ReadLine() ) != null )
            {
                string[] tokens = line.Split( new Char[] { ' ' } );

                if ( tokens.Length < 6 )
                {
                    success = false;

                    break;
                }

                // Recording coordinates
                coordinates_.Add( Convert.ToSingle( tokens[ 0 ] ) );
                coordinates_.Add( Convert.ToSingle( tokens[ 1 ] ) );
                coordinates_.Add( Convert.ToSingle( tokens[ 2 ] ) );

                // Recording normals
                normals_.Add( Convert.ToSingle( tokens[ 3 ] ) );
                normals_.Add( Convert.ToSingle( tokens[ 4 ] ) );
                normals_.Add( Convert.ToSingle( tokens[ 5 ] ) );
            }

            reader.Close();

            return success;
        }
    }
}